core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -995,6 +1003,19 @@ private[spark] class BlockManager(
     None
   }
 
+  private[spark] def getHostLocalDirs(executorIds: Array[String])
+      : scala.collection.Map[String, Array[String]] = {
```

Comment:
The predef `Map` is the immutable one. With just using `Map` as return type I would got:
```
Error:(1012, 19) type mismatch;
 found   : Map[String,Array[String]] (in scala.collection) 
 required: Map[String,Array[String]] (in scala.collection.immutable) 
      cachedItems ++ notCachedItems
```` 

So I decided to avoid unnecessary conversion (copy of items)  I would use the the `Map` trait. See `toMap` implementation:

```
def toMap[T, U](implicit ev: A <:< (T, U)): immutable.Map[T, U] = {
    val b = immutable.Map.newBuilder[T, U]
    for (x <- self)
      b += x

    b.result()
  }
```

Reply:
Thanks! It is very great catch! 

I have double checked and the `synchronized` is really need as tasks (i.e. `ShuffleMapTask`) are executed in different threads because of `org.apache.spark.executor.Executor#launchTask`.